### DISCUSSION TITLE
Only one remaining issue before the first official release : what next ?

### DISCUSSION THREAD

**Discussion by Rodemfr** [Rodemfr/MicronetToNMEA#40] [Ideas]
Today, there is only one remaining issue for MicronetToNMEA : LSM303DLHC axes issue. I ordered a board to fix this but I can expect it to be solve next week. Once solved, I will build the first officiel release of MicronetToNMEA.
Now comes the question : what next ?

I plan to continue maintenance of the project for as long as it will be needed, however, I forsee that porting MTN to ESP32 is something waited by potential users. ESP32, appart from beeing unexpensive, allows to easily enable a WiFi hotspot for data broadcasting. Also, after discussing with @frgaude, I discovered interesting boards like this one : [ESP32](http://www.lilygo.cn/prod_view.aspx?TypeId=50060&Id=1317&FId=t3:50060:3)

This is not a CC1101, but I read the datasheet of the SX127x and I understand that we can disable the Lora demodulator to use a standard FSK one. This would be quite some work but if successful, we would have a simple 40-60€ board with all of what is needed to MTN : Micronet RF, GNSS, Bluetooth & Wifi. Only LSM303 would need to be added. This looks appealing.

**Comment by frgaude** [comment]
Hi @Rodemfr,

I am indeed quite interested in participating in a port to the ESP32 platform. These boards are easier to find than the Teensy ones.

The Lilygo T-Beam boards do pack modules and features directly useful for the project, in a very small form-factor.  And to add to your list, they can run on a 18650 Li-Ion battery, they usually come with a small (1-inch) OLED display and there are some nice 3d-printed cases available.

But which board variant to use as they come in a variety of different GNSS and different radios?
- **NEO-6M** (GPS only) or **NEO-M8N** (slightly better accuracy, refresh rate, GPS & GLONASS)
- **SX1276** or **SX1262** (newer),  for 868Mhz/915Mhz radio

On the GNSS side, the **M8N** apparently shows slightly better accuracy and could potentially connect to Galileo* satelites
_*:if firmware-upgradable and if it can be flashed while soldered to the board. There are many "fake" M8N on chinese boards that can't be updated, I will try to check the module that comes with the T-Beam._
-> A **6M** is probably sufficent.

On the radio side, the newer **SX1262** radio is  advertised as more powerful while using less current when receiving ... but it has a totally different set of commands, so it would require a specific driver.
-> it is worth aligning on which radio to use, and **SX1276** is the more common one.

Also I was surprised by how small the GNSS antenna is. I would replace it with a regular size one (25mm square) for boat use.

The price difference from 6M-SX1276 to M8N-SX1262 is around 13 EUR/USD (with the mini oled screen).

Let us know your thoughts on the choice of GNSS & radio.

**Comment by Rodemfr** [comment]
I ordered a LILYGO T-beam 1.1 for 868MHz. It is a NEO-6M GNSS but I don't know which RF IC I have 1262 or 127x ? I will check tonight.
I ordered this one to start evaluating porting of MicronetToNMEA but it is not a final choice.
I also have noticed the [LILYGO T-ECHO](https://ae01.alicdn.com/kf/H6b077bbc8b624467887db060275c5f23R.jpg).
I must say I love the nice box with buttons and a e-ink display and the processor (M4@64Mhz) is less power hungry than ESP32. But I see two problems : 
- if you want to connect your NMEA stream with something else than Bluetooth or USB, you need to open the box, make holes and solder wires.
- Powering it from boat's battery will likely require some DC/DC downconverter somewhere.

**Reply by frgaude** [reply]
You will get an SX1276 on this board. and I got one with an SX1262 :-)

These T-ECHO have a really small screen. From a display standpoint, I'm more interested in the T5 series with their 4.7inch epaper displays (there are also 2.13'' and 7.5'' variants). T5 has no LoRa chip.

Apparently for outdoor use, the screen sensitivity to UV light is a key issue. (interesting long thread [here](http://www.parapentiste.info/forum/developpements-hardware-software/diy-gnuvario-variometre-opensource-openhardware-arduino-t48334.0.html) ).
so either a dedicated screen or some UV protection film is needed.

**Comment by frgaude** [comment]
here are some ideas I 'd be interested in exploring:

(already mentioned)
- ESP32 port, providing integrated BLE and WiFI bridges 
- SX12xx radio support

(then in no particular order)
- more options to configure the ublox GNNS devices
- maybe a web interface to complement the nice serial one (for when we only have a phone and MTN)
- implement a master mode for MTN, which would help when no Tacktick display devices are on the network (asumming T121 or T122 don't handle master mode but i don't know).
- allow MTN to awake the devices on the Micronet network
- DIY Micronet displays
   * epaper, perfect for daylight + solution for night (frontlighting or lcd screen)
   * should it connect to Micronet wireless or be a more generic Wifi UDP display?      
- embedded small oled display like on the tbeam boards (for basic info)
- DIY Micronet remote control
    * my pet project 
- connecting Seatalk / serial NMEA / CAN bus NMEA / other serial devices to MTN boards ( probably more of a documentaion effort than a coding effort)
    * auto pilot
    * voltage monitor
    * existing NMEA or Seatalk sensors ...
- with a LoRa/Micronet radio on the board, we could envision to turn our boats into wind stations. To get wind readings from home, thanks to LoRa long range / LoRaWan :-)

**Comment by Rodemfr** [comment]
I received my T-Beam. You are right its a SX1276, but if your board has the same connections and only the SX1262 as a difference, it should not be too hard to make a dedicated RF driver. I ported MicronetToNMEA on it but, of course, CC1101 and LSM303 drivers are disabled. It's now time to look at the existing drivers for SX127x...

The Lilygo T5 is really eye-catching, I would love to have it on top of my chart table !

Thanks for the feature list. I like your proposal for configuring GNSS and to have a web interface for configuration.

**Reply by frgaude** [reply]
Wow, you are really impressive, the port is already on the way! Kudos!

**Comment by Rodemfr** [comment]
Porting is on going but it will take time before a first alpha version, probably one or two months. Both esp32/FreeRTOS environment and SX1276 require deep changes in MicronetToNMEA architecture.
The good news is that I think SX1276 is perfectly fitted for Micronet. So is probably SX1262.

**Comment by Rodemfr** [comment]
Porting work on TBEAM 1.1 is progressing : 
 - EEPROM OK
 - GNSS OK
 - **Micronet RX OK !!**

Now I'm working on Micronet TX...

**Reply by frgaude** [reply]
Amazing news!!

On my end, I was toying around with the MTN source code and various esp32s (regular / C3), PlatformIO, JTAG debugging, etc, but I get stuck a lot. still fun and I'm hopefully learning a bit.

I'm eager to test your code! and I will need to try porting your SX1276 code to the SX1262 (!).

By the way, did you notice that you could plug the LSM303 pin to pin on the T-Beam, (using a female pin header block on the pins they advertise for the tiny oled display) ? it really makes a very compact setup! 

Keep up the good work!

**Comment by dwarning** [comment]
I got just the t-beam v1.1 and made first experiments to come familiar with this amazing part.
I tried to catch MTN messages from display but no luck so far with this code:
`#include <RadioLib.h>

#define RADIO_CS_PIN                18
#define RADIO_DIO0_PIN              26
#define RADIO_RST_PIN               23
#define RADIO_DIO1_PIN              33
#define RADIO_BUSY_PIN              32
SX1276 radio = new Module(RADIO_CS_PIN, RADIO_DIO0_PIN, RADIO_RST_PIN, RADIO_BUSY_PIN);

void setup() {
  Serial.begin(9600);

  // initialize SX1276 with default settings
  Serial.print(F("[SX1276] Initializing ... "));
  int state = radio.beginFSK();

  if (state == RADIOLIB_ERR_NONE) {
    Serial.println(F("success!"));
  } else {
    Serial.print(F("failed, code "));
    Serial.println(state);
    while (true);
  }
  state = radio.setFrequency(869.84);
  state = radio.setBitRate(76.8);
  state = radio.setFrequencyDeviation(38.0);
  state = radio.setRxBandwidth(250);
  state = radio.setGain(0); // automatic
  state = radio.setDataShaping(RADIOLIB_SHAPING_NONE);
  state = radio.disableAddressFiltering(); 
  state = radio.setPreambleLength(15);
  uint8_t syncWord[] = {0x55, 0x99};
  state = radio.setSyncWord(syncWord, 2);
  if (state != RADIOLIB_ERR_NONE) {
    Serial.print(F("Unable to set configuration, code "));
    Serial.println(state);
    while (true);
  }

}

void loop() {
  Serial.println(F("[SX1276] Waiting for incoming transmission ... "));

  byte byteArr[8];
  int state = radio.receive(byteArr, 8);

  if (state == RADIOLIB_ERR_NONE) {
    // packet was successfully received
    Serial.println(F("success!"));

    // print the data of the packet
    Serial.print(F("[SX1276] Data:\t\t\t"));
    for (int i = 0; i<8; i++){
      Serial.print(byteArr[i],HEX);
      Serial.print(" ");
    }
    Serial.println();

    // print the RSSI (Received Signal Strength Indicator)
    // of the last received packet
    Serial.print(F("[SX1276] RSSI:\t\t\t"));
    Serial.print(radio.getRSSI());
    Serial.println(F(" dBm"));

    // print the SNR (Signal-to-Noise Ratio)
    // of the last received packet
    Serial.print(F("[SX1276] SNR:\t\t\t"));
    Serial.print(radio.getSNR());
    Serial.println(F(" dB"));

    // print frequency error
    // of the last received packet
    Serial.print(F("[SX1276] Frequency error:\t"));
    Serial.print(radio.getFrequencyError());
    Serial.println(F(" Hz"));

  } else if (state == RADIOLIB_ERR_RX_TIMEOUT) {
    // timeout occurred while waiting for a packet
    Serial.println(F("timeout!"));

  } else if (state == RADIOLIB_ERR_CRC_MISMATCH) {
    // packet was received, but is malformed
    Serial.println(F("CRC error!"));

  } else {
    // some other error occurred
    Serial.print(F("failed, code "));
    Serial.println(state);
  }

}
`
But no luck so far, also with a version in using interrupt on DIO0 pin.
@Rodemfr : can you give few hints how you handled this.

**Comment by dwarning** [comment]
I got the rx working, with Radiolib and FIFO operation.

**Comment by Rodemfr** [comment]
I added you to my private repository with the port of MicronetToNMEA to T-Beam v1.1.
This port requires the T-Beam v1.1 with SX1276 and the OLED display. I hope you ordered this one. If not, the display can be ordered independently : a 128*64 display driven by SSD1306.
I entirely rewrote the SX1276 driver, using RadioLib as a reference.
It works well in both RX & TX.
I want to first release a fully working version of MicronetToNMEA before making the T-Beam port public.
